sources/host-ctr/cmd/host-ctr/pid_namespaces.go

```diff
@@ -0,0 +1,172 @@
+	pid := (*task).Pid()
+	pidNsFile, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+	if err != nil {
+		return -1, fmt.Errorf("failed to open %d PID namespace file", pid)
+	}
+	pidNsFd := int(pidNsFile.Fd())
+
+	// Get file descriptor that refers to a parent namespace, ioctl() request '_IO(NSIO, 0x2)'
+	_, err = unix.IoctlGetInt(pidNsFd, unix.NS_GET_PARENT)
+	if err == unix.EPERM {
+		// If we can't access the parent PID namespace, this means the current container task is sharing the root PID
+		// namespace. We don't want to track the root PID namespace in the manifest, so we return -1.
+		return -1, nil
+	} else if err != nil {
+		return -1, errors.Wrap(err, fmt.Sprintf("ioctl failed to 'NS_GET_PARENTNS' for '%s'", pidNsFile.Name()))
+	}
+
+	// The PID namespace is represented by the inode number of the task's PID namespace file
+	var sb unix.Stat_t
+	if err := unix.Fstat(pidNsFd, &sb); err != nil {
+		return -1, errors.Wrap(err, "failed to stat task PID namespace fd")
+	}
+
+	return int64(sb.Ino), nil
+}
+
+// HostCtrPidNs represents the list of PID namespaces
+type HostCtrPidNs struct {
+	PidNamespaces []int64 `json:"pid-namespaces,omitempty"`
+}
+
+// writeManifest writes the list of host-container PID NS to the open manifest file
+func writeManifest(hostCtrPidNsList *HostCtrPidNs, manifest *os.File) error {
+	newData, err := json.Marshal(*hostCtrPidNsList)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal PID NS list to JSON")
+	}
+	if err := manifest.Truncate(0); err != nil {
+		return errors.Wrap(err, "failed to truncate PID NS manifest")
+	}
+	if _, err := manifest.Seek(0, 0); err != nil {
+		return errors.Wrap(err, "failed to reset PID NS manifest I/O offset")
+	}
+	if _, err = manifest.Write(newData); err != nil {
+		return errors.Wrap(err, "failed to write to PID NS manifest")
+	}
+	return nil
+}
+
+func addPidNsToManifest(pidNs int64) (err error) {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	locked, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	if !locked {
+		return errors.New("failed to acquire lock for writing to PID NS manifest")
+	}
+	defer func() {
+		err = fileLock.Unlock()
+	}()
+
+	// Add task PID NS to manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	var (
+		tmpPidNsList     HostCtrPidNs
+		hostCtrPidNsList HostCtrPidNs
+	)
```

Comment:
Nit: typically, I've seen the community prefer the `:=` operator. Something like:

```go
tmpPidNsList := &HostCtrPidNs{}
hostCtrPidNsList := &HostCtrPidNs{}
```

This is a bit more clear that you're allocating empty `HostCtrPidNs`